File: Physics/2. Electricity/Project/project.py
import numpy as np
import matplotlib.pyplot as plt

# Константы и параметры
k = 1  # Условная константа кулоновской силы
charge_magnitude = 1  # Модуль заряда (в безразмерной форме)
grid_size = 100 
grid_range = 2.0 
epsilon = 1e-2  # Точность для завершения итераций
max_steps = 5  # Максимальное количество шагов

# Координаты и значения зарядов
charges = [
    {"x": -0.5, "y": -0.5, "q": charge_magnitude},
    {"x": 0.5, "y": -0.5, "q": -charge_magnitude},
    {"x": -0.5, "y": 0.5, "q": -charge_magnitude},
    {"x": 0.5, "y": 0.5, "q": charge_magnitude},
]

x = np.linspace(-grid_range, grid_range, grid_size)
y = np.linspace(-grid_range, grid_range, grid_size)
X, Y = np.meshgrid(x, y)

# Вычисление потенциала (эквипотенциальные поверхности)
phi = np.zeros(X.shape)
for charge in charges:
    dx = X - charge["x"]
    dy = Y - charge["y"]
    r = np.sqrt(dx**2 + dy**2)
    r[r < 0.1] = 0.1
    phi += k * charge["q"] / r

for step in range(max_steps):
    phi_old = phi.copy()
    # Обновление значений потенциала
    for i in range(1, grid_size - 1):
        for j in range(1, grid_size - 1):
            if not any(np.isclose([X[i, j], Y[i, j]], [charge["x"], charge["y"]], atol=0.1).all() for charge in charges):
                phi[i, j] = 0.25 * (phi[i+1, j] + phi[i-1, j] + phi[i, j+1] + phi[i, j-1])
    diff = np.max(np.abs(phi - phi_old))
    print(f"Шаг {step + 1}, изменение потенциала: {diff:.4e}")
    if diff < epsilon:
        print(f"Сходимость достигнута на шаге {step + 1}")
        break
else:
    print("Максимальное количество шагов достигнуто.")

# Вычисление напряженности поля
Ex = np.zeros(X.shape)
Ey = np.zeros(Y.shape)
for charge in charges:
    dx = X - charge["x"]
    dy = Y - charge["y"]
    r = np.sqrt(dx**2 + dy**2)
    r[r < 0.1] = 0.1  # Предотвращение деления на 0
    Ex += k * charge["q"] * dx / r**3
    Ey += k * charge["q"] * dy / r**3

E_magnitude = np.sqrt(Ex**2 + Ey**2)
Ex_normalized = Ex / E_magnitude
Ey_normalized = Ey / E_magnitude

plt.figure(figsize=(8, 6))
plt.contour(X, Y, phi, levels=50, cmap="coolwarm", alpha=0.75)
plt.streamplot(X, Y, Ex_normalized, Ey_normalized, color=E_magnitude, cmap="viridis", density=1)
for charge in charges:
    color = 'red' if charge["q"] > 0 else 'blue'
    plt.plot(charge["x"], charge["y"], 'o', color=color, markersize=10)
plt.title("Силовые линии и эквипотенциальные поверхности")
plt.xlabel("x (безразмерные координаты)")
plt.ylabel("y (безразмерные координаты)")
plt.colorbar(label="Потенциал / Напряженность")
plt.axis("equal")
plt.grid(True)
plt.show()

File: Physics/2. Electricity/Project/project.m
% Параметры зарядов
charges = [1, -1, 1, -1]; % Заряды (Кл)
charges_x = [-1, 1, 1, -1]; % Координаты зарядов по x (м)
charges_y = [-1, -1, 1, 1]; % Координаты зарядов по y (м)

% Сетка расчета (уменьшена для ускорения)
x = linspace(-5, 5, 100);
y = linspace(-5, 5, 100);
[X, Y] = meshgrid(x, y);

% Константы
k = 9e9; % Электростатическая постоянная (Н·м²/Кл²)
num_frames = 200; % Количество кадров анимации
dt = 0.05; % Шаг времени для движения зарядов
speed = 20; % Скорость движения зарядов (м/с) — увеличена для более быстрого движения

figure;

for frame = 1:num_frames
    % Обновление координат зарядов (хаотичное движение)
    charges_x = charges_x + (rand(1, length(charges)) - 0.5) * speed * dt;
    charges_y = charges_y + (rand(1, length(charges)) - 0.5) * speed * dt;
    charges_x = max(min(charges_x, max(x)), min(x));
    charges_y = max(min(charges_y, max(y)), min(y));

    % Расчёт потенциала
    phi = zeros(size(X)); % Матрица для потенциала
    for i = 1:length(charges)
        r = sqrt((X - charges_x(i)).^2 + (Y - charges_y(i)).^2);
        phi = phi + k * charges(i) ./ r;
    end
    phi(isinf(phi)) = NaN;

    % Масштабирование потенциала для визуализации
    phi_scaled = log(abs(phi));

    % Расчёт напряжённости электрического поля (градиент потенциала)
    [Ex, Ey] = gradient(-phi, x(2) - x(1), y(2) - y(1));

    % Построение эквипотенциальных поверхностей
    subplot(1, 2, 1);
    contour(X, Y, phi_scaled, 50, 'LineWidth', 1.5);
    hold on;
    scatter(charges_x(charges > 0), charges_y(charges > 0), 100, 'filled', 'r');
    scatter(charges_x(charges < 0), charges_y(charges < 0), 100, 'filled', 'b');
    hold off;
    axis equal;
    title('Эквипотенциальные поверхности');
    xlabel('x, м');
    ylabel('y, м');
    colormap('parula');
    colorbar;

    % Построение силовых линий
    subplot(1, 2, 2);
    quiver(X, Y, Ex, Ey, 'k');
    hold on;
    streamslice(X, Y, Ex, Ey)
    scatter(charges_x(charges > 0), charges_y(charges > 0), 100, 'filled', 'r');
    scatter(charges_x(charges < 0), charges_y(charges < 0), 100, 'filled', 'b');
    hold off;
    axis equal;
    title('Силовые линии');
    xlabel('x, м');
    ylabel('y, м');

    pause(1 / 60);
    drawnow;
end

disp('Анимация завершена.');
